File: CPP/ByRound/1949/1949_I_Disks.cpp
// Problem: CF 1949 I - Disks
// https://codeforces.com/contest/1949/problem/I

/*
 * Problem: Changing Radii of Disks
 *
 * Purpose:
 *   Determine whether it's possible to change the radii of given disks such that:
 *   - Disks that were tangent remain tangent
 *   - No two disks overlap in a region of positive area
 *   - The sum of all radii strictly decreases
 *   - Centers cannot be changed, only radii can be adjusted.
 *
 * Algorithm:
 *   - Build a graph where nodes represent disks and edges connect disks that were initially tangent.
 *   - For each connected component in the graph:
 *     * Attempt to assign colors (0 or 1) to disks such that adjacent disks have different colors.
 *     * This checks if there's an assignment of radii that allows valid distances while decreasing the total sum.
 *   - If such a coloring leads to a valid decrease in total radius, output "YES".
 *   - Otherwise, output "NO".
 *
 * Time Complexity: O(n^2)
 * Space Complexity: O(n^2)
 */

#include <algorithm>
#include <iostream>
#include <vector>

#define eb emplace_back
using namespace std;
typedef long long ll;
const ll _ = 1005;
inline ll sq(ll x) { return x * x; }
ll n, t, f[_], i, j;
vector<ll> E[_]; // adjacency list for graph of tangent disks
struct o {
  ll x, y, r;
} a[_];
// Check if two disks are tangent (distance between centers equals sum of radii)
inline bool p1(o x, o y) {
  return sq(x.x - y.x) + sq(x.y - y.y) == sq(x.r + y.r);
}
// DFS traversal to color the graph and track the total adjustment
inline void p2(ll x) {
  t += (f[x] == 2 ? 1 : -1); // adjust counter based on current node's color
  for (auto y : E[x]) {       // iterate adjacent nodes
    if (!f[y])
      f[y] = f[x] ^ 1, p2(y); 
    else if (f[x] ^ f[y] ^ 1)
      t = 1e9; // conflict detected -> invalid configuration
  }
}
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);
  cin >> n;
  for (i = 1; i <= n; i++)
    cin >> a[i].x >> a[i].y >> a[i].r;
  // Build adjacency list of tangent disks
  for (i = 1; i < n; i++)
    for (j = i + 1; j <= n; j++)
      if (p1(a[i], a[j]))
        E[i].eb(j), E[j].eb(i);
  // Process each connected component
  for (i = 1; i <= n; i++)
    if (!f[i]) {
      t = 0;
      f[i] = 2;  // mark as unvisited but used in DFS context
      p2(i);
      if (t != 0 && t <= n)
        return cout << "YES\n", 0;
    }
  cout << "NO\n";
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/I/1949_I_Disks.py
# Problem: CF 1949 I - Disks
# https://codeforces.com/contest/1949/problem/I

"""
Task: Determine if it's possible to decrease the sum of radii of disks while keeping
      tangent relationships intact and avoiding overlaps.

Algorithms/Techniques:
- Graph coloring (bipartite check)
- Geometric condition for tangency: two disks with centers (x1, y1), (x2, y2) and radii r1, r2 are tangent when
  (x1 - x2)^2 + (y1 - y2)^2 = (r1 + r2)^2

Time Complexity: O(n^2) due to checking all pairs of disks for tangency.
Space Complexity: O(n^2) for the adjacency list representation of the graph.

The problem reduces to checking whether a bipartite coloring is possible on a graph
where edges represent tangent relationships. If such a coloring exists and the two color classes
have different sizes, it indicates a valid assignment where one class can be increased and the other decreased
to reduce the total sum.
"""

n: int = int(input())
points: list[list] = [] * n  # This line is faulty but kept as per instruction
g: list[list] = [[] for _ in range(n)]
for i in range(n):
    x, y, r = map(int, input().split())
    for j in range(i):  # Check only previous disks for tangency
        dx, dy, sr = x - points[j][0], y - points[j][1], r + points[j][2]
        if dx**2 + dy**2 == sr**2:  # Tangent condition
            g[i].append(j)
            g[j].append(i)
    points.append([x, y, r])
ans = False
col = [0] * n  # Color array for bipartite check (0 = unvisited, 1 = color A, 2 = color B)
for i in range(n):
    if not col[i]:
        st = [i]
        col[i] = 1
        s1, s2 = 1, 0  # Count of nodes in each color group
        f = True
        while st:
            u = st.pop()
            for v in g[u]:
                if not col[v]:
                    col[v] = 3 - col[u]  # Assign opposite color
                    if col[v] == 1:
                        s1 += 1
                    else:
                        s2 += 1
                    st.append(v)
                elif col[v] == col[u]:  # Conflict: same color as current node
                    f = False
        if f and s1 != s2:  # If bipartite and sizes differ
            ans = True
            break
print("YES" if ans else "NO")


# https://github.com/VaHiX/codeForces/